### PR TITLE
[WIP] Fix issue #4975

### PR DESCRIPTION
**Pull request by theodoretliu**
Addresses issue #4975. See three test cases, first is passing, unclear on what the behavior should be on second two

### CODE REVIEW

**Comment by theodoretliu** [comment]
@ilevkivskyi @Michael0x2a appears to be working if I hardcode a file `test.py` and run `mypy` on it
```python
from typing import Tuple

x: Tuple[str, float]
y: Tuple[int, ...]

lst = [x, y]
reveal_type(lst[0])
```
produces
```
test.py:7: note: Revealed type is 'builtins.tuple*[builtins.object]'
```
with no errors.

However, running the same in the test suite seems to cause issues.

Similar with the following
```python
from typing import Tuple

x: Tuple[int, str]
y: Tuple[int, ...]

lst = [x, y]
reveal_type(lst[0])
```
produces
```
test.py:7: note: Revealed type is 'builtins.tuple*[builtins.object]'
```

File: mypy/join.py [hunk 1]
```diff
@@ -268,6 +270,12 @@ def visit_tuple_type(self, t: TupleType) -> ProperType:
                                       mypy.typeops.tuple_fallback(t))
             assert isinstance(fallback, Instance)
             return TupleType(items, fallback)
+        elif isinstance(self.s, Instance) and self.s.type.fullname == 'builtins.tuple':
```
Comment (ilevkivskyi) [inline mypy/join.py]: Couple more things:
* What if the instance is a _subclass_ of `tuple`?
* What if we have a named tuple on either side?

Comment (theodoretliu) [reply mypy/join.py]: I pushed some updates to try and handle the subclass case. I'm afraid I don't really understand the subclassing system. I know to look in .type.mro to see the hierarchy of types, but was unclear how to get the original "args" of the `builtins.tuple` type. For example
```python
class A(Tuple[int, ...]):
    pass

class B(A):
    pass

x: B
y: Tuple[int]
```
How do we access the `builtins.int` part when joining?

Comment (theodoretliu) [reply mypy/join.py]: Once this is resolved, I can work on the NamedTuple case as well

**Comment by ilevkivskyi** [comment]
Sorry, I will not have time for this, hopefully @JukkaL or @msullivan can help.

**Comment by JukkaL** [comment]
I can finish the review. The next steps would be fixing the merge conflict and the failing test cases. If you need help with some of these, just let me know.

**Comment by intgr** [comment]
Doh, sorry :( I was under the impression that nobody was working on this issue (made that assumption when JelleZijlstra commented in #4975). It didn't help that this PR is titled "Initial attempt"

So meanwhile I submitted two pull requests for the two behaviors implemented here: #8333 and #8335.

The good news is that I happened to find a cleaner implementation, my approach can join `Tuple[x, x] + List[x]` into a `Sequence[x]` type, which seems to not be the case here.

I now tested my code with the test cases in this PR and my code passes these cases.

What do you think is the best way forward?

**Comment by theodoretliu** [comment]
@intgr well, you were quick on the uptake, so I don't have a problem if they merge yours instead of mine. In the future, I can title my PR better, but I did reference the issue in this PR and it does show up on the Issue tracker... Nevertheless, taking a quick look at your PR shows that we had approximately the same idea, it's up the maintainers @JukkaL @ilevkivskyi to make the final call.

There was an issue with my PR that was raised regarding types that inherit from Tuple: if we go with yours, we might try to fix that too?

**Comment by intgr** [comment]
Thanks, for the heads-up, I hadn't considered Tuple subclasses. I'll have to test that behavior.

**Comment by JukkaL** [comment]
Hmm, I think I prefer #8333 and #8335, since they are more recent and have no conflicts. Supporting tuple subclasses would be a nice addition.

**Comment by theodoretliu** [comment]
Closing in favor of @intgr work in #8333 and #8335

**Pull request closed without merging**